lib/ts/Diags.cc

```diff
@@ -226,43 +214,46 @@ void
 Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocation *loc, const char *format_string,
                 va_list ap) const
 {
-  struct timeval tp;
-  const char *s;
-  char *buffer, *d, timestamp_buf[48];
-  char format_buf[1024], format_buf_w_ts[1024], *end_of_format;
+  const size_t BUF_CAPACITY = 1024;
+  using ts::LocalBufferWriter;
 
   ink_release_assert(diags_level < DiagsLevel_Count);
 
-  ////////////////////////////////////////////////////////////////////////
-  // there are 2 format buffers that hold a printf-style format string  //
-  // format_buf contains <prefix_string>: (<debug_tag>) <format_string> //
-  // and format_buf_w_ts has the same thing with a prepended timestamp. //
-  ////////////////////////////////////////////////////////////////////////
+  LocalBufferWriter<BUF_CAPACITY> format_writer;
+
+  // Save room for optional newline and terminating NUL bytes.
+  format_writer.clip(2);
 
-  format_buf[0]      = NUL;
-  format_buf_w_ts[0] = NUL;
+  {
+    struct timeval tp = ink_gettimeofday();
+    time_t cur_clock = (time_t)tp.tv_sec;
+    char timestamp_buf[48];
+    char *buffer = ink_ctime_r(&cur_clock, timestamp_buf);
 
-  /////////////////////////////////////////////////////
-  // format_buf holds 1024 characters, end_of_format //
-  // points to the current available character       //
-  /////////////////////////////////////////////////////
+    snprintf(&(timestamp_buf[19]), (sizeof(timestamp_buf) - 20), ".%03d", (int)(tp.tv_usec / 1000));
 
-  end_of_format  = format_buf;
-  *end_of_format = NUL;
+    format_writer.write('[');
+    format_writer.write(buffer + 4, std::min(strlen(buffer + 4), sizeof(timestamp_buf) - 4));
+    format_writer.write("] ", 2);
+  }
 
-  // add the thread id
-  end_of_format += snprintf(end_of_format, sizeof(format_buf), "{0x%" PRIx64 "} ", (uint64_t)ink_thread_self());
+  size_t timestamp_end_offset = format_writer.extent();
+
+  {
+    // add the thread id
+    char tmp[BUF_CAPACITY];
+    int num_bytes_written = snprintf(tmp, format_writer.remaining(), "{0x%" PRIx64 "} ", (uint64_t)ink_thread_self());
+    if (num_bytes_written > 0) {
+      format_writer.write(tmp, num_bytes_written);
+    }
+  }
 
   //////////////////////////////////////
   // start with the diag level prefix //
   //////////////////////////////////////
 
-  for (s = level_name(diags_level); *s; *end_of_format++ = *s++) {
-    ;
-  }
-
-  *end_of_format++ = ':';
-  *end_of_format++ = ' ';
+  format_writer.write(level_name(diags_level), strlen(level_name(diags_level)));
+  format_writer << ": ";
```

Comment:
Should use `write` here for consistency and to avoid constructing a string_view.